example-code/snippets/src/backend/Fonts.test.ts

```diff
@@ -0,0 +1,122 @@
+    async function getSystemFontFamilies(includeNonEmbeddable = false): Promise<Map<string, FontFile[]>> {
+      // Map font family names to the font file(s) containing their faces.
+      const families = new Map<string, FontFile[]>();
+      // Query the absolute filename for every available system font.
+      const systemFontPaths = await getSystemFonts();
+      // Add the font families from each font file to the map.
+      for (const systemFontPath of systemFontPaths) {
+        let fontFile;
+        try {
+          fontFile = FontFile.createFromTrueTypeFileName(systemFontPath);
+          if (!includeNonEmbeddable && !fontFile.isEmbeddable) {
+            continue;
+          }
+        } catch (_) { // eslint-disable-line @typescript-eslint/no-unused-vars
+          continue;
+        }
+
+        for (const face of fontFile.faces) {
+          let list = families.get(face.familyName);
+          if (!list) {
+            families.set(face.familyName, list = []);
+          }
+
+          list.push(fontFile);
+        }
+      }
+
+      return families;
+    }
+    // __PUBLISH_EXTRACT_END__
+
+    const sysFams = await getSystemFontFamilies();
+    expect(sysFams.size).greaterThan(0);
+    
+    // __PUBLISH_EXTRACT_START__ Fonts.selectSystemFont
+    async function selectAndEmbedSystemFont(db: IModelDb): Promise<FontId | undefined> {
+      // Query available font families.
+      const availableFamilies = await getSystemFontFamilies();
+      // Ask the user to select the family to be installed.
+      const familyToInstall = await askUserToChooseFontFamilyToInstall(availableFamilies.keys());
+      if (!familyToInstall || !availableFamilies.get(familyToInstall)) {
+        // No family selected to install.
+        return undefined;
+      }
+
+      const descriptor: FontFamilyDescriptor = {
+        name: familyToInstall,
+        type: FontType.TrueType,
+      };
+
+      // If the family's faces are already embedded and the family is already assigned an Id, we're finished.
+      for (const family of db.fonts.queryMappedFamilies()) {
+        if (family.name === descriptor.name && family.type === descriptor.type) {
+          return family.id;
+        }
+      }
+
+      // Install all of the font files containing the selected families faces.
+      const filesToInstall = availableFamilies.get(familyToInstall)!;
+      await Promise.all(filesToInstall.map(async (file) => db.fonts.embedFontFile({ file })));
+      return db.fonts.findId(descriptor);
+    }
+    // __PUBLISH_EXTRACT_END__
+
+    const sysFontId = await selectAndEmbedSystemFont(iModel);
+    expect(sysFontId).to.equal(1);
+  });
+
+  it("embeds user-provided SHX font", async () => {
+    // __PUBLISH_EXTRACT_START__ Fonts.embedShxFont
+    async function embedShxFont(filename: string, db: IModelDb): Promise<FontId> {
+      // Read the contents of the SHX file into memory.
+      const blob = fs.readFileSync(filename);
```

Comment:
Since the embedShxFont function is async, would it be better to use the async readFile function from fs/promises? Note: I'm honestly not sure if doing so is considered better or not. It is non-blocking, but I'm not sure what else can happen while embedShxFont is running.